File: labs/09-mcp-server-stdio/python/main.py
"""Minimal MCP server over stdio (JSON-RPC with length-prefixed framing).

Implements:
  - initialize
  - notifications/initialized
  - tools/list
  - tools/call

Tools:
  - get_current_time(timezone: str = "UTC") -> ISO timestamp
  - calculate(expression: str) -> number
"""

import argparse
import ast
import json
import logging
import operator
import sys
from datetime import datetime, timezone as _timezone
from typing import Any

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    stream=sys.stderr,
)
logger = logging.getLogger("mcp-server")


# ---------- safe calculator ----------

_ALLOWED_BIN_OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
    ast.Pow: operator.pow,
    ast.Mod: operator.mod,
}

_ALLOWED_UNARY_OPS = {
    ast.UAdd: operator.pos,
    ast.USub: operator.neg,
}


def safe_eval(node: ast.AST) -> float:
    if isinstance(node, ast.Expression):
        return safe_eval(node.body)
    if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
        return float(node.value)
    if isinstance(node, ast.BinOp):
        op_type = type(node.op)
        if op_type not in _ALLOWED_BIN_OPS:
            raise ValueError(f"Unsupported binary operator: {op_type.__name__}")
        return _ALLOWED_BIN_OPS[op_type](safe_eval(node.left), safe_eval(node.right))
    if isinstance(node, ast.UnaryOp):
        op_type = type(node.op)
        if op_type not in _ALLOWED_UNARY_OPS:
            raise ValueError(f"Unsupported unary operator: {op_type.__name__}")
        return _ALLOWED_UNARY_OPS[op_type](safe_eval(node.operand))
    raise ValueError(f"Unsupported expression: {type(node).__name__}")


def calculate(expression: str) -> dict[str, Any]:
    try:
        tree = ast.parse(expression.strip(), mode="eval")
        value = safe_eval(tree)
        return {"content": [{"type": "text", "text": str(value)}], "isError": False}
    except Exception as exc:
        return {
            "content": [{"type": "text", "text": f"Invalid expression: {exc}"}],
            "isError": True,
        }


# ---------- tools ----------

TOOLS = [
    {
        "name": "get_current_time",
        "description": "Return the current time in ISO 8601 format.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "timezone": {
                    "type": "string",
                    "description": "IANA timezone name, e.g. UTC or Asia/Shanghai.",
                }
            },
        },
    },
    {
        "name": "calculate",
        "description": "Evaluate a simple arithmetic expression.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "expression": {
                    "type": "string",
                    "description": "Arithmetic expression with +, -, *, /, parentheses, numbers.",
                }
            },
            "required": ["expression"],
        },
    },
]


def get_current_time(timezone: str = "UTC") -> dict[str, Any]:
    try:
        import zoneinfo

        tz = zoneinfo.ZoneInfo(timezone)
    except Exception:
        tz = _timezone.utc if timezone.upper() == "UTC" else None
    if tz is None:
        return {
            "content": [
                {
                    "type": "text",
                    "text": f"Unsupported timezone: {timezone}. Using UTC.",
                }
            ],
            "isError": True,
        }
    now = datetime.now(tz)
    return {
        "content": [{"type": "text", "text": now.isoformat()}],
        "isError": False,
    }


TOOL_HANDLERS = {
    "get_current_time": get_current_time,
    "calculate": calculate,
}


# ---------- JSON-RPC framing ----------


def read_message(stream: Any) -> dict[str, Any] | None:
    """Read a length-prefixed JSON-RPC message from a binary stream."""
    headers = {}
    while True:
        line = stream.readline()
        if not line:
            return None
        line = line.decode("utf-8").strip()
        if not line:
            break
        if ":" in line:
            key, value = line.split(":", 1)
            headers[key.strip().lower()] = value.strip()

    length = int(headers.get("content-length", 0))
    if length == 0:
        return None
    body = stream.read(length)
    return json.loads(body.decode("utf-8"))


def write_message(stream: Any, message: dict[str, Any]) -> None:
    """Write a length-prefixed JSON-RPC message to a binary stream."""
    body = json.dumps(message, ensure_ascii=False).encode("utf-8")
    stream.write(f"Content-Length: {len(body)}\r\n\r\n".encode("utf-8"))
    stream.write(body)
    stream.flush()


# ---------- request handlers ----------


def make_response(request_id: Any, result: Any) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "result": result}


def make_error(request_id: Any, code: int, message: str) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "error": {"code": code, "message": message}}


def handle_request(request: dict[str, Any]) -> dict[str, Any] | None:
    method = request.get("method")
    params = request.get("params", {})
    req_id = request.get("id")

    logger.info("Received %s", method)

    if method == "initialize":
        return make_response(
            req_id,
            {
                "protocolVersion": "2024-11-05",
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "agent-labs-mcp-server", "version": "0.1.0"},
            },
        )

    if method == "notifications/initialized":
        return None  # notification, no response

    if method == "tools/list":
        return make_response(req_id, {"tools": TOOLS})

    if method == "tools/call":
        name = params.get("name")
        arguments = params.get("arguments", {})
        handler = TOOL_HANDLERS.get(name)
        if not handler:
            return make_error(req_id, -32601, f"Unknown tool: {name}")
        try:
            result = handler(**arguments)
        except Exception as exc:
            return make_error(req_id, -32603, f"Tool error: {exc}")
        return make_response(req_id, result)

    return make_error(req_id, -32601, f"Method not found: {method}")


def serve(stdin=None, stdout=None) -> None:
    stdin = stdin or sys.stdin.buffer
    stdout = stdout or sys.stdout.buffer
    logger.info("MCP server ready on stdio")
    while True:
        try:
            request = read_message(stdin)
        except json.JSONDecodeError as exc:
            logger.error("Bad JSON: %s", exc)
            write_message(stdout, make_error(None, -32700, "Parse error"))
            continue
        if request is None:
            logger.info("EOF reached; shutting down")
            break

        response = handle_request(request)
        if response is not None:
            write_message(stdout, response)


# ---------- smoke test ----------


def smoke_test():
    """Run a self-contained smoke test using an in-memory stdin/stdout pair."""
    import io

    requests = [
        {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "initialize",
            "params": {"protocolVersion": "2024-11-05", "capabilities": {}},
        },
        {"jsonrpc": "2.0", "id": None, "method": "notifications/initialized"},
        {"jsonrpc": "2.0", "id": 2, "method": "tools/list"},
        {
            "jsonrpc": "2.0",
            "id": 3,
            "method": "tools/call",
            "params": {"name": "get_current_time", "arguments": {"timezone": "UTC"}},
        },
        {
            "jsonrpc": "2.0",
            "id": 4,
            "method": "tools/call",
            "params": {"name": "calculate", "arguments": {"expression": "(2 + 3) * 4"}},
        },
    ]

    stdin = io.BytesIO()
    for req in requests:
        body = json.dumps(req).encode("utf-8")
        stdin.write(f"Content-Length: {len(body)}\r\n\r\n".encode("utf-8"))
        stdin.write(body)
    stdin.seek(0)

    stdout = io.BytesIO()
    serve(stdin, stdout)
    stdout.seek(0)
    print("Smoke test output:")
    print(stdout.read().decode("utf-8"))


def main():
    parser = argparse.ArgumentParser(description="Minimal MCP server over stdio")
    parser.add_argument(
        "--smoke",
        action="store_true",
        help="Run a self-contained smoke test instead of serving.",
    )
    args = parser.parse_args()
    if args.smoke:
        smoke_test()
    else:
        serve()


if __name__ == "__main__":
    main()

File: labs/09-mcp-server-stdio/go/main.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

var logger = log.New(os.Stderr, "mcp-server: ", log.LstdFlags)

// ---------- safe calculator ----------

type calcParser struct {
	s string
	i int
}

func (p *calcParser) peek() byte {
	if p.i >= len(p.s) {
		return 0
	}
	return p.s[p.i]
}

func (p *calcParser) next() byte {
	ch := p.peek()
	if p.i < len(p.s) {
		p.i++
	}
	return ch
}

func (p *calcParser) skipSpaces() {
	for p.peek() == ' ' {
		p.next()
	}
}

func safeEvalExpression(s string) (float64, error) {
	p := &calcParser{s: s}
	return p.parseExpr()
}

func (p *calcParser) parseExpr() (float64, error) {
	left, err := p.parseTerm()
	if err != nil {
		return 0, err
	}
	for {
		p.skipSpaces()
		ch := p.peek()
		if ch != '+' && ch != '-' {
			break
		}
		p.next()
		right, err := p.parseTerm()
		if err != nil {
			return 0, err
		}
		if ch == '+' {
			left += right
		} else {
			left -= right
		}
	}
	return left, nil
}

func (p *calcParser) parseTerm() (float64, error) {
	left, err := p.parseFactor()
	if err != nil {
		return 0, err
	}
	for {
		p.skipSpaces()
		ch := p.peek()
		if ch != '*' && ch != '/' && ch != '%' {
			break
		}
		p.next()
		right, err := p.parseFactor()
		if err != nil {
			return 0, err
		}
		switch ch {
		case '*':
			left *= right
		case '/':
			if right == 0 {
				return 0, fmt.Errorf("division by zero")
			}
			left /= right
		case '%':
			left = float64(int64(left) % int64(right))
		}
	}
	return left, nil
}

func (p *calcParser) parseFactor() (float64, error) {
	p.skipSpaces()
	ch := p.peek()
	if ch == '(' {
		p.next()
		val, err := p.parseExpr()
		if err != nil {
			return 0, err
		}
		p.skipSpaces()
		if p.peek() != ')' {
			return 0, fmt.Errorf("expected ')'")
		}
		p.next()
		return val, nil
	}
	if ch == '+' || ch == '-' {
		p.next()
		val, err := p.parseFactor()
		if err != nil {
			return 0, err
		}
		if ch == '-' {
			return -val, nil
		}
		return val, nil
	}
	return p.parseNumber()
}

func (p *calcParser) parseNumber() (float64, error) {
	p.skipSpaces()
	start := p.i
	hasDot := false
	for {
		ch := p.peek()
		if ch >= '0' && ch <= '9' {
			p.next()
		} else if ch == '.' && !hasDot {
			hasDot = true
			p.next()
		} else {
			break
		}
	}
	if start == p.i {
		return 0, fmt.Errorf("expected number")
	}
	return strconv.ParseFloat(p.s[start:p.i], 64)
}

func calculate(expression string) map[string]interface{} {
	val, err := safeEvalExpression(strings.TrimSpace(expression))
	if err != nil {
		return map[string]interface{}{
			"content": []map[string]string{{"type": "text", "text": fmt.Sprintf("Invalid expression: %v", err)}},
			"isError": true,
		}
	}
	return map[string]interface{}{
		"content": []map[string]string{{"type": "text", "text": fmt.Sprintf("%v", val)}},
		"isError": false,
	}
}

var tools = []map[string]interface{}{
	{
		"name":        "get_current_time",
		"description": "Return the current time in ISO 8601 format.",
		"inputSchema": map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"timezone": map[string]interface{}{
					"type":        "string",
					"description": "IANA timezone name, e.g. UTC or Asia/Shanghai.",
				},
			},
		},
	},
	{
		"name":        "calculate",
		"description": "Evaluate a simple arithmetic expression.",
		"inputSchema": map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"expression": map[string]interface{}{
					"type":        "string",
					"description": "Arithmetic expression with +, -, *, /, parentheses, numbers.",
				},
			},
			"required": []string{"expression"},
		},
	},
}

var toolHandlers = map[string]func(map[string]interface{}) map[string]interface{}{
	"get_current_time": func(args map[string]interface{}) map[string]interface{} {
		tzName := "UTC"
		if v, ok := args["timezone"].(string); ok && v != "" {
			tzName = v
		}
		loc, err := time.LoadLocation(tzName)
		if err != nil {
			return map[string]interface{}{
				"content": []map[string]string{{"type": "text", "text": fmt.Sprintf("Unsupported timezone: %s. Using UTC.", tzName)}},
				"isError": true,
			}
		}
		return map[string]interface{}{
			"content": []map[string]string{{"type": "text", "text": time.Now().In(loc).Format(time.RFC3339)}},
			"isError": false,
		}
	},
	"calculate": func(args map[string]interface{}) map[string]interface{} {
		expr, _ := args["expression"].(string)
		return calculate(expr)
	},
}

// ---------- JSON-RPC framing ----------

func readMessage(r *bufio.Reader) (map[string]interface{}, error) {
	headers := map[string]string{}
	for {
		line, err := r.ReadString('\n')
		if err != nil {
			if err == io.EOF && len(headers) == 0 {
				return nil, io.EOF
			}
			return nil, err
		}
		line = strings.TrimSpace(line)
		if line == "" {
			break
		}
		if idx := strings.Index(line, ":"); idx >= 0 {
			key := strings.ToLower(strings.TrimSpace(line[:idx]))
			value := strings.TrimSpace(line[idx+1:])
			headers[key] = value
		}
	}
	length, _ := strconv.Atoi(headers["content-length"])
	if length == 0 {
		return nil, nil
	}
	body := make([]byte, length)
	_, err := io.ReadFull(r, body)
	if err != nil {
		return nil, err
	}
	var msg map[string]interface{}
	if err := json.Unmarshal(body, &msg); err != nil {
		return nil, err
	}
	return msg, nil
}

func writeMessage(w io.Writer, msg map[string]interface{}) error {
	body, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	_, err = fmt.Fprintf(w, "Content-Length: %d\r\n\r\n%s", len(body), body)
	return err
}

// ---------- request handlers ----------

func makeResponse(requestID interface{}, result interface{}) map[string]interface{} {
	return map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      requestID,
		"result":  result,
	}
}

func makeError(requestID interface{}, code int, message string) map[string]interface{} {
	return map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      requestID,
		"error":   map[string]interface{}{"code": code, "message": message},
	}
}

func handleRequest(request map[string]interface{}) map[string]interface{} {
	method, _ := request["method"].(string)
	params, _ := request["params"].(map[string]interface{})
	reqID := request["id"]

	logger.Printf("Received %s", method)

	switch method {
	case "initialize":
		return makeResponse(reqID, map[string]interface{}{
			"protocolVersion": "2024-11-05",
			"capabilities":    map[string]interface{}{"tools": map[string]interface{}{}},
			"serverInfo":      map[string]interface{}{"name": "agent-labs-mcp-server", "version": "0.1.0"},
		})
	case "notifications/initialized":
		return nil
	case "tools/list":
		return makeResponse(reqID, map[string]interface{}{"tools": tools})
	case "tools/call":
		name, _ := params["name"].(string)
		arguments, _ := params["arguments"].(map[string]interface{})
		handler := toolHandlers[name]
		if handler == nil {
			return makeError(reqID, -32601, fmt.Sprintf("Unknown tool: %s", name))
		}
		return makeResponse(reqID, handler(arguments))
	default:
		return makeError(reqID, -32601, fmt.Sprintf("Method not found: %s", method))
	}
}

func serve(stdin io.Reader, stdout io.Writer) {
	r := bufio.NewReader(stdin)
	w := bufio.NewWriter(stdout)
	logger.Println("MCP server ready on stdio")
	for {
		request, err := readMessage(r)
		if err != nil {
			if err == io.EOF {
				logger.Println("EOF reached; shutting down")
				break
			}
			logger.Printf("Bad JSON: %v", err)
			_ = writeMessage(w, makeError(nil, -32700, "Parse error"))
			_ = w.Flush()
			continue
		}
		if request == nil {
			logger.Println("EOF reached; shutting down")
			break
		}
		response := handleRequest(request)
		if response != nil {
			_ = writeMessage(w, response)
			_ = w.Flush()
		}
	}
}

// ---------- smoke test ----------

func smokeTest() {
	requests := []map[string]interface{}{
		{
			"jsonrpc": "2.0",
			"id":      1,
			"method":  "initialize",
			"params":  map[string]interface{}{"protocolVersion": "2024-11-05", "capabilities": map[string]interface{}{}},
		},
		{"jsonrpc": "2.0", "id": nil, "method": "notifications/initialized"},
		{"jsonrpc": "2.0", "id": 2, "method": "tools/list"},
		{
			"jsonrpc": "2.0",
			"id":      3,
			"method":  "tools/call",
			"params":  map[string]interface{}{"name": "get_current_time", "arguments": map[string]interface{}{"timezone": "UTC"}},
		},
		{
			"jsonrpc": "2.0",
			"id":      4,
			"method":  "tools/call",
			"params":  map[string]interface{}{"name": "calculate", "arguments": map[string]interface{}{"expression": "(2 + 3) * 4"}},
		},
	}

	var stdin bytes.Buffer
	for _, req := range requests {
		body, _ := json.Marshal(req)
		stdin.WriteString(fmt.Sprintf("Content-Length: %d\r\n\r\n%s", len(body), body))
	}

	var stdout bytes.Buffer
	serve(&stdin, &stdout)
	fmt.Println("Smoke test output:")
	fmt.Println(stdout.String())
}

func main() {
	smoke := flag.Bool("smoke", false, "Run a self-contained smoke test instead of serving.")
	flag.Parse()
	if *smoke {
		smokeTest()
	} else {
		serve(os.Stdin, os.Stdout)
	}
}
